File: test/ModularPipelines.UnitTests/Execution/NotInParallelTestsWithConstraintKeys.cs
using ModularPipelines.TestHelpers;

namespace ModularPipelines.UnitTests.Execution;

[TUnit.Core.NotInParallel]
public class NotInParallelTestsWithConstraintKeys : TestBase
{
    private static readonly NotInParallelTracker Tracker = new();

    [ModularPipelines.Attributes.NotInParallel("A")]
    public class ModuleWithAConstraintKey1 : NotInParallelTestModule
    {
        protected override NotInParallelTracker Tracker => NotInParallelTestsWithConstraintKeys.Tracker;
        protected override IEnumerable<string> ConflictingModuleNames => ["ModuleWithAConstraintKey2"];
    }

    [ModularPipelines.Attributes.NotInParallel("A")]
    public class ModuleWithAConstraintKey2 : NotInParallelTestModule
    {
        protected override NotInParallelTracker Tracker => NotInParallelTestsWithConstraintKeys.Tracker;
        protected override IEnumerable<string> ConflictingModuleNames => ["ModuleWithAConstraintKey1"];
    }

    [ModularPipelines.Attributes.NotInParallel("B")]
    public class ModuleWithBConstraintKey1 : NotInParallelTestModule
    {
        protected override NotInParallelTracker Tracker => NotInParallelTestsWithConstraintKeys.Tracker;
        protected override IEnumerable<string> ConflictingModuleNames => ["ModuleWithBConstraintKey2"];
    }

    [ModularPipelines.Attributes.NotInParallel("B")]
    public class ModuleWithBConstraintKey2 : NotInParallelTestModule
    {
        protected override NotInParallelTracker Tracker => NotInParallelTestsWithConstraintKeys.Tracker;
        protected override IEnumerable<string> ConflictingModuleNames => ["ModuleWithBConstraintKey1"];
    }

    [Test]
    public async Task NotInParallel_If_Same_ConstraintKey()
    {
        Tracker.Reset();

        await TestPipelineHostBuilder.Create()
            .AddModule<ModuleWithAConstraintKey1>()
            .AddModule<ModuleWithAConstraintKey2>()
            .AddModule<ModuleWithBConstraintKey1>()
            .AddModule<ModuleWithBConstraintKey2>()
            .ExecutePipelineAsync();

        await Assert.That(Tracker.Violations).IsEmpty();
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Execution/NotInParallelTestsWithConstraintKeys.fs
namespace ModularPipelines.UnitTests.FSharp.Execution

open System.Collections.Generic
open ModularPipelines.Extensions
open ModularPipelines.TestHelpers
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core

module private NotInParallelTestsWithConstraintKeysData =
    let tracker = NotInParallelTracker()

[<ModularPipelines.Attributes.NotInParallel("A")>]
type ModuleWithAConstraintKey1() =
    inherit NotInParallelTestModule()
    override _.Tracker = NotInParallelTestsWithConstraintKeysData.tracker
    override _.ConflictingModuleNames = ["ModuleWithAConstraintKey2"] :> IEnumerable<string>

[<ModularPipelines.Attributes.NotInParallel("A")>]
type ModuleWithAConstraintKey2() =
    inherit NotInParallelTestModule()
    override _.Tracker = NotInParallelTestsWithConstraintKeysData.tracker
    override _.ConflictingModuleNames = ["ModuleWithAConstraintKey1"] :> IEnumerable<string>

[<ModularPipelines.Attributes.NotInParallel("B")>]
type ModuleWithBConstraintKey1() =
    inherit NotInParallelTestModule()
    override _.Tracker = NotInParallelTestsWithConstraintKeysData.tracker
    override _.ConflictingModuleNames = ["ModuleWithBConstraintKey2"] :> IEnumerable<string>

[<ModularPipelines.Attributes.NotInParallel("B")>]
type ModuleWithBConstraintKey2() =
    inherit NotInParallelTestModule()
    override _.Tracker = NotInParallelTestsWithConstraintKeysData.tracker
    override _.ConflictingModuleNames = ["ModuleWithBConstraintKey1"] :> IEnumerable<string>

type NotInParallelTestsWithConstraintKeys() =
    inherit TestBase()

    [<Test>]
    member _.NotInParallel_If_Same_ConstraintKey() = async {
        NotInParallelTestsWithConstraintKeysData.tracker.Reset()

        let! _ = TestPipelineHostBuilder.Create()
                    .AddModule<ModuleWithAConstraintKey1>()
                    .AddModule<ModuleWithAConstraintKey2>()
                    .AddModule<ModuleWithBConstraintKey1>()
                    .AddModule<ModuleWithBConstraintKey2>()
                    .ExecutePipelineAsync()
                |> Async.AwaitTask

        do! check(Assert.That<string>(NotInParallelTestsWithConstraintKeysData.tracker.Violations).IsEmpty())
    }
